rln/src/poseidon_tree.rs

```diff
@@ -36,16 +49,101 @@ impl Hasher for PoseidonHash {
     }
 }
 
+// The pmtree Hasher trait used by pmtree Merkle tree
+impl pmtree::Hasher for PoseidonHash {
+    type Fr = Fr;
+
+    fn default_leaf() -> Self::Fr {
+        Fr::from(0)
+    }
+
+    fn serialize(value: Self::Fr) -> Value {
+        fr_to_bytes_le(&value)
+    }
+
+    fn deserialize(value: Value) -> Self::Fr {
+        let (fr, _) = bytes_le_to_fr(&value);
+        //let c = std::io::Cursor::new(value);
+        //Fr::deserialize(c).unwrap()
+        fr
+    }
+
+    fn hash(inputs: &[Self::Fr]) -> Self::Fr {
+        poseidon_hash(inputs)
+    }
+}
+
+// pmtree supports in-memory and on-disk databases (Database trait) for storing the Merkle tree state
+
+// We implement Database for hashmaps, an in-memory database
+struct MemoryDB(HashMap<DBKey, Value>);
+
+impl Database for MemoryDB {
+    fn new(_dbpath: &str) -> Result<Self> {
+        Ok(MemoryDB(HashMap::new()))
+    }
+
+    fn load(_dbpath: &str) -> Result<Self> {
+        Err(Error("Cannot load in-memory DB".to_string()))
+    }
+
+    fn get(&self, key: DBKey) -> Result<Option<Value>> {
+        Ok(self.0.get(&key).cloned())
+    }
+
+    fn put(&mut self, key: DBKey, value: Value) -> Result<()> {
+        self.0.insert(key, value);
+
+        Ok(())
+    }
+}
+
+// We implement Database for sled DB, an on-disk database
+struct SledDB(Sled);
+
+impl Database for SledDB {
+    fn new(dbpath: &str) -> Result<Self> {
+        if Path::new(dbpath).exists() {
+            fs::remove_dir_all(dbpath).unwrap();
+        }
+        let db = sled::open(dbpath).unwrap();
+        Ok(SledDB(db))
+    }
+
+    fn load(dbpath: &str) -> Result<Self> {
+        let db = sled::open(dbpath).unwrap();
+
+        if !db.was_recovered() {
+            return Err(Error("Trying to load non-existing database!".to_string()));
+        }
+
+        Ok(SledDB(db))
+    }
+
+    fn get(&self, key: DBKey) -> Result<Option<Value>> {
+        Ok(self.0.get(key).unwrap().map(|val| val.to_vec()))
+    }
+
+    fn put(&mut self, key: DBKey, value: Value) -> Result<()> {
+        self.0.insert(key, value).unwrap();
```

Comment:
We should return the error here, not unwrap
```suggestion
        self.0.insert(key, value)?;
```
Or even better, have an enum for errors so that they can be appropriately handled by users.

Reply:
That's what I was talking about before, but atm I don't know how to make simple & clear error handling. The problem is in `Error` compatibility: pmtree Database trait methods return custom result (defined in pmtree) and Sled::Db return their own errors.

Need to discuss and decide, maybe I can reimplement something in pmtree, so error handling will be better.

Reply:
Absolutely, let's go with the first method :)

Reply:
Probably late to the dance. But unless targeting no-std, you could probably use thiserror crate instead.